components/compliance-service/reporting/relaxting/reports.go

```diff
@@ -530,6 +499,87 @@ func (backend *ES2Backend) GetReport(esIndex string, reportId string,
 	return report, errorutils.ProcessNotFound(nil, reportId)
 }
 
+func convertControl(profileControlsMap map[string]*reportingapi.Control, reportControlMin ESInSpecReportControl, filters map[string][]string) *reportingapi.Control {
+	profileControl := profileControlsMap[reportControlMin.ID]
+
+	if profileControl != nil {
+		profileControl.Results = make([]*reportingapi.Result, 0)
+		if reportControlMin.ID == profileControl.Id {
+			minResults := make([]*reportingapi.Result, len(reportControlMin.Results))
+			for i, result := range reportControlMin.Results {
+				minResults[i] = &reportingapi.Result{
+					Status:      result.Status,
+					CodeDesc:    result.CodeDesc,
+					RunTime:     result.RunTime,
+					Message:     result.Message,
+					SkipMessage: result.SkipMessage,
+				}
+			}
+
+			convertedControl := reportingapi.Control{
+				Id:             profileControl.Id,
+				Code:           profileControl.Code,
+				Desc:           profileControl.Desc,
+				Impact:         profileControl.Impact,
+				Title:          profileControl.Title,
+				SourceLocation: profileControl.SourceLocation,
+				Results:        minResults,
+			}
+
+			jsonTags := make(map[string]*reportingapi.TagValues, 0)
+			for _, tag := range reportControlMin.StringTags {
+				if len(tag.Values) == 0 {
+					jsonTags[tag.Key] = &reportingapi.TagValues{Values: []string{"null"}}
+				}
+				vals := make([]string, 0)
+				for _, val := range tag.Values {
+					vals = append(vals, val)
+					jsonTags[tag.Key] = &reportingapi.TagValues{Values: vals}
+				}
+			}
+			matchingTagFound := false
+			matchingTagRequired := false
+			for filterKey, filterVals := range filters {
+				if strings.HasPrefix(filterKey, "control_tag") {
+					matchingTagRequired = true
+					trimmed := strings.TrimPrefix(filterKey, "control_tag:")
+					if tagVal, ok := jsonTags[trimmed]; ok {
+						for _, val := range filterVals {
+							if contains(tagVal.Values, val) || val == "null" {
+								matchingTagFound = true
+								break
+							}
+						}
+					}
+
+				}
+			}
```

Comment:
This could be moved to its own function like below. 
```
func doesControlTagMatchFilter(filters map[string][]string,
	jsonTags map[string]*reportingapi.TagValues) bool {
	for filterKey, filterVals := range filters {
		if !strings.HasPrefix(filterKey, "control_tag") {
			return true
		}
		trimmed := strings.TrimPrefix(filterKey, "control_tag:")
		if tagVal, ok := jsonTags[trimmed]; ok {
			for _, val := range filterVals {
				if contains(tagVal.Values, val) || val == "null" {
					return true
				}
			}
		}
	}
	return false
}
```
Then in the function you would have the below
```
	if !doesControlTagMatchFilter(filters, jsonTags) {
		return nil
	}
```

Reply:
👍